File: admin-frontend/src/types/backend/dto/system/MenuDto.ts
interface MenuBase {
  id: string
  parentId: string | null
  level: number
  menuType: 'directory' | 'menu' | 'action'
  menuCode: string
  actionCode: string | null
  menuName: string
  menuPath: string | null
  sortOrder: number
  canEdit: boolean
  isHide: boolean
}

interface MenuDirectory extends MenuBase {
  menuType: 'directory'
  actionCode: null
  menuPath: null
}

interface MenuMenu extends MenuBase {
  menuType: 'menu'
  actionCode: null
  menuPath: string
}

interface MenuAction extends MenuBase {
  menuType: 'action'
  actionCode: string
  menuPath: null
}

export type MenuDto =
  | MenuDirectory
  | MenuMenu
  | MenuAction

File: admin-backend/src/main/java/com/example/backend/modules/system/model/dto/MenuDto.java
package com.example.backend.modules.system.model.dto;

import lombok.Data;

import java.io.Serializable;
import java.util.List;

@Data
public class MenuDto implements Serializable {

    private String id;
    private String parentId;
    private Integer level;
    private String menuType;
    private String menuCode;
    private String actionCode;
    private String menuName;
    private String menuPath;
    private Integer sortOrder;
    private Boolean canEdit;
    private Boolean isHide;
    private List<MenuDto> children;

}
